parser/src/transforms/inlining.rs

```diff
@@ -293,9 +302,14 @@ impl<'a> Inlining<'a> {
             }
             // Expression statements are introduced during inlining, and are always already expanded,
             // but they are recursively visited to apply rewrites
-            Statement::Expr(mut expr) => {
-                self.rewrite_expr(&mut expr)?;
-                Ok(vec![Statement::Expr(expr)])
+            Statement::Expr(mut expr) => match expr {
+                Expr::Call(call) => {
```

Comment:
I would revert this specific change here - when we expand a statement block, we will expand call sites, and in the process recursively expand the callee. We do want to rewrite call expressions which are introduced into the tree when inlining the callee (so that arguments are rewritten to use the expressions from the caller). I think expanding the call here would cause inlining of the callee to occur without rewriting the arguments, and result in incorrect rewriting of the callee body.

I'm working through this by running the code in my mental interpreter, so to speak, so I may be forgetting something that makes this necessary - but the comment I left just above this is what reminded me that this is already supposed to be an expression introduced as the result of inlining, and we only are interested in applying rewrites.

It may also be very possible that `Expr::Call` can never be observed here - a good way to check that would be to make the match arm panic and see if you can trigger it. Without working through some tests designed to see what exactly is going on when a call is seen here, it's hard for me to say for sure whether that would be unexpected with certainty, but my gut feeling is that we don't want to attempt expanding calls here.

Reply:
Question: Would we not need it when the return value is a function call of another function?
```
fn foo(a: felt) -> felt:
    return bar(a)
```

Maybe there's another way to handle this?

Reply:
Expansion is performed recursively, so when we expand the call to `foo`, we will also expand it's body, and thus the call to `bar` as well.

As an aside, we need to make sure we check for recursive functions calls and raise a diagnostic. We can do that during inlining by tracking what calls we're currently expanding. Each time we start expanding a call, we check if we're currently in an expansion of the callee, and if so, raise a diagnostic; if not, then we insert the callee into the visited set and proceed with inlining. When we finish expanding a call, we remove the callee from the set.

We should probably also do that for evaluators, since the same issue can occur there, I don't think I handled that anywhere yet.

We can handle that later though, no need to do that as part of this PR.